### PR TITLE
[GEOT-5409] GroupByVisitor does not allow collection wrappers to optimize visit

### PR DESCRIPTION
**Pull request by aaime**

### CODE REVIEW

File: modules/library/main/src/test/java/org/geotools/feature/visitor/GroupByVisitorTest.java [hunk 1]
```diff
@@ -282,6 +291,21 @@ public void testMergingGroupByResults() throws Exception {
                 new Object[]{"THEATER", 200.0}
         });
     }
+    
+    @Test
+    public void testFeatureAttributeVisitor() {
+        GroupByVisitor visitor = buildVisitor("energy_consumption", "Average", new String[]{"building_type"});
+        List<Expression> expressions = visitor.getExpressions();
+        Set<String> names = new HashSet<String>();
```
Comment (nmco) [inline modules/library/main/src/test/java/org/geotools/feature/visitor/GroupByVisitorTest.java]: new HashSet<String>() could be new HashSet<>()

File: modules/library/main/src/test/java/org/geotools/feature/visitor/GroupByVisitorTest.java [hunk 2]
```diff
@@ -305,14 +329,21 @@ private void testVisitor(String aggregateAttribute, String aggregateVisitor,
...
+            String[] groupByAttributes) {
         GroupByVisitorBuilder visitorBuilder = new GroupByVisitorBuilder()
                 .withAggregateAttribute(aggregateAttribute, buildingType)
                 .withAggregateVisitor(aggregateVisitor);
         for (String groupByAttribute : groupByAttributes) {
             visitorBuilder.withGroupByAttribute(groupByAttribute, buildingType);
         }
         GroupByVisitor visitor = visitorBuilder.build();
-        featureCollection.accepts(visitor, new NullProgressListener());
         return visitor;
```
Comment (nmco) [inline modules/library/main/src/test/java/org/geotools/feature/visitor/GroupByVisitorTest.java]: The visitor variable is not need could simply be: `return visitorBuilder.build();`

File: modules/library/main/src/main/java/org/geotools/feature/visitor/GroupByVisitor.java [hunk 3]
```diff
@@ -337,4 +343,11 @@ public Map toMap() {
             return result;
         }
     }
+
+    @Override
+    public List<Expression> getExpressions() {
+        List<Expression> result = new ArrayList<Expression>(groupByAttributes);
```
Comment (nmco) [inline modules/library/main/src/main/java/org/geotools/feature/visitor/GroupByVisitor.java]: Type argument Expression is not need

File: modules/library/main/src/test/java/org/geotools/feature/visitor/GroupByVisitorTest.java (hunk 2 above)
Comment (aaime) [inline modules/library/main/src/test/java/org/geotools/feature/visitor/GroupByVisitorTest.java]: We often do this on purpose to simplify debugging (gives a place where the variable can be examined before exiting the method), but sure, in this case it adds no value

**Comment by nmco** [comment]
Tried it on a real system and the visitor execution is correctly delegated. Sorry for the missed interface :)

**Comment by aaime** [comment]
Applied Nuno's feedback, merging

**Pull request merged**